### PR TITLE
[VarDumper] Add date caster

### PR DESCRIPTION
**Pull request by maidmaid**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | /
| License       | MIT
| Doc PR        | /

I propose to add a ``DateCaster`` that casts date (with timestamp, date, time, timezone (offset + id), literal date, delta from now and DST) and interval.

Todo:
- [x] cast date
- [x] cast interval
- [x] add tests

### CODE REVIEW

**Comment by ro0NL** [comment]
Btw.. about the timezone. Maybe put the offset in front, this is the part i usually care about;

```
timezone: '+02:00 (Europe/Amsterdam)'
```

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    public static function castDate(DateTimeInterface $d, $a, Stub $stub, $isNested, $filter)
+    {
+        $prefix = Caster::PREFIX_VIRTUAL;
+
+        return array(
+            $prefix.'date' => $d->format('Y-m-d H:i:s.u'),
+            $prefix.'literal' => $d->format('l, j M Y'),
+            $prefix.'timestamp' => $d->getTimestamp(),
+            $prefix.'timezone' => $d->format('P (e)'),
+            $prefix.'Δnow' => (new DateTime())->diff($d)->format('%R %yy %mm %dd %H:%I:%S'),
```
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Needs a `setTimezone($d->getTimezone())` right?

Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Good catch!

**Comment by maidmaid** [comment]
screenshot from 2017-04-14 22-30-46
WDYT of this variant?

**Comment by ro0NL** [comment]
Not sure.. it looks a bit more chaotic imo.

**Comment by maidmaid** [comment]
``literal`` and ``timestamp`` are not visible if they are nested or if we flag with ``EXCLUDE_VERBOSE``. I have too add tests.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+use DateTime;
+use DateTimeInterface;
+use Symfony\Component\VarDumper\Cloner\Stub;
+
+/**
+ * Casts DateTimeInterface related classes to array representation.
+ *
+ * @author Dany Maillard <danymaillard93b@gmail.com>
+ */
+class DateCaster
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: should be named `DateTimeCaster` IMO

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\VarDumper\Caster;
+
+use DateTime;
+use DateTimeInterface;
```
Comment (stof) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: We never add use statements for global classes

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by robfrawley** [comment]
@maidmaid I was a fan of the first proposal. The unix time is almost always the most important field to me (though this could depend on the operating system one uses, I guess).

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 2 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: In a next PR, I would cast other date classes like ``DateInterval``. It's why a generic ``DateCaster`` is better IMO.

**Comment by maidmaid** [comment]
@robfrawley Yes, unix time can be usefull, but it's a machin format and we should favor human representation, IMO.

**Comment by ogizanagi** [comment]
@maidmaid : It depends on the situation, but you might be happy to have the unix time while debugging. And I'm mainly using the var-dumper component for debugging purpose 😄

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 3 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Removed

**Comment by maidmaid** [comment]
We have lots of different interesting opinions. @nicolas-grekas, can you give your opinion?

**Comment by nicolas-grekas** [comment]
May I propose something different:
always display a single line that represents 100% of the state of the objects: `'Y-m-d H:i:s.u e'`
then lets provide more details on hover (... from now, literal, etc as a single "title" line).
This is doable with a single virtual propery holding a ConstStub I think.

**Comment by maidmaid** [comment]
> then lets provide more details on hover

But works it with ``CliDumper``?

**Comment by nicolas-grekas** [comment]
CliDumper will display only `Y-m-d H:i:s.u e`, but this LGTM, that's still 100% of the state.

**Comment by nicolas-grekas** [comment]
For the reader: when I saw that, I first thought "hey, this is hiding the actual internal structure of DateTime objects" (by not telling about these existing public props.)
The I ran this: https://3v4l.org/uWc54
I looks like doing var_dump/array-cast actually *creates* the 3 props...
So I'm even more in favor of doing something here, and I would add DateTimeZone, etc. to the list of classes that should display only virtual props.

**Comment by maidmaid** [comment]
When we var_dump a ``DateTime`` object, we can see 3 props in all but 2 distinct props: date and timezone. So, ``DateCaster`` should cast a ``date`` prop that represents 100% of date/time infos (with more details on hover), and a ``timezone`` prop that shows a ``DateTimeZone`` thanks to an other caster (that is already in WIP [here](https://github.com/maidmaid/symfony/compare/master...datecaster-datezone#diff-959db3511f6175d0172b859e897e62dbR113)). Right?

**Comment by nicolas-grekas** [comment]
worth a try yes :)

**Comment by maidmaid** [comment]
As discussed:

screenshot from 2017-05-02 13-40-00

Too bad it is not possible to copy timestamp and to see extra info in CLI with this version. More, I find it strange to use ``ConstStub`` to shows extra in hover, it seems as a hack.

**Comment by nicolas-grekas** [comment]
Personally, I like compact formats, that give me everything in a snap.
I would also remove the timezone prop in fact, and add the "e" modifier at the end of the date string.
Nesting is an opportunity to be hit by depth limitings, better not.
Copy/pasting the timestamp looks like something I wouldn't do very often to me (especially since it's destructive compared to a DateTime object).
Of course, that's only my personal PoV :)

**Comment by nicolas-grekas** [comment]
What about displaying the timestamp could be displayed next to the class name `DateTime @123456789 { ... }`? Should be just a matter of appending it to the `$stub->class` prop.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 4]
```diff
@@ -25,14 +25,16 @@ public static function castDate(\DateTimeInterface $d, array $a, Stub $stub, $is
...
-            $a[$prefix.'literal'] = $d->format('l, j F Y');
-            $a[$prefix.'timestamp'] = $d->getTimestamp();
-        }
+        $a[$prefix.'date'] = new ConstStub(
+            $d->format('Y-m-d H:i:s.u'),
+            sprintf(
+                "literal: %s\ntimestamp: %s\nΔnow: %s",
+                $d->format('l, j F Y'),
+                $d->getTimestamp(),
+                (new \DateTime('now', $d->getTimezone()))->diff($d)->format('%R %yy %mm %dd %H:%I:%S')
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: `$d->getTimezone()` is not required in this context

**Comment by maidmaid** [comment]
I putted timestamp in ``$stub->class`` and display timezone in ``date`` virtual prop (used ``P`` instead ``e`` because ``+02:00`` is more informative that ``Europe/Zurich``):

screenshot from 2017-05-02 15-56-31

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 4 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: Removed

**Comment by nicolas-grekas** [comment]
Do you like it? I think I do :)

> used P instead e because +02:00 is more informative

in fact, that's not true: `Europe/Zurich` contains geographical/political info that makes it very different from `+02:00`.
See http://stackoverflow.com/questions/17694894/different-timezone-types-on-datetime-object

**Comment by maidmaid** [comment]
Yes, I find this version quite elegant :)

About timezone, I think that we have to know difference to GMT when we debug a date but not its geographical/political infos. Are you agree?

My only little remark rests on ``ConstStub`` hack usage to showing extra infos when ``[ConstStub] Represents a PHP constant and its value``.

**Comment by nicolas-grekas** [comment]
> know difference to GMT when we debug a date but not its geographical/political infos

but this prevents knowing if the timezone of a DateTime is of type 3 or not. And since this is of significant importance (see SO), I think we should use "e".

About ConstStub, I'd say we don't really care :)

**Comment by maidmaid** [comment]
What about to have ``P (e)`` format, like ``+02:00 (Europe/Zurich)``? We have diff to GMT + geo info + type; everybody is happy like that :smile:

**Comment by nicolas-grekas** [comment]
Not sure:
```php
$a = new \DateTime('now');
$a->setTimezone(new DateTimezone('GMT+2'));
echo $a->format('P e');
// output: +02:00 +02:00
```

**Comment by maidmaid** [comment]
This kind of notation (``GMT+n`` and others) is not recommanded by PHP; only the ``Europe/Zurich``-like is the good way. See warnings here http://php.net/manual/en/timezones.others.php. Following these recommendations, ``P (e)`` format displays fine all the time.

**Comment by nicolas-grekas** [comment]
But we don't code for the ideal case only :)

**Comment by robfrawley** [comment]
Can we not display `P (e)` when they are different and only `e` when the are the same value, thereby providing a more robust result if the `DateTime` object is a "type 3" and not showing duplicate values when it is a "type 1" or "type 2" (re [SO@17694894](http://stackoverflow.com/questions/17694894/different-timezone-types-on-datetime-object)).

```php
<?php

function timezoneToString(\DateTime $dateTime)
{
    $p = $dateTime->format('P');
    $e = $dateTime->format('e');

    return $e === $p ? $e : sprintf('%s (%s)', $p, $e);
}

$date = new \DateTime('now', new DateTimezone('GMT+2'));
echo sprintf("@%d: timezone %s\n", $date->format('U'), timezoneToString($date));

$date = new \DateTime('now', new \DateTimezone('America/New_York'));
echo sprintf("@%d: timezone %s\n", $date->format('U'), timezoneToString($date));
```

Results (the important part being after the `timezone` string, obviously):

```
@1493746514: timezone +02:00
@1493746514: timezone -04:00 (America/New_York)
```

**Comment by ogizanagi** [comment]
What about https://3v4l.org/uqc19 ?

**Comment by robfrawley** [comment]
@ogizanagi That looks great to me!

**Comment by ro0NL** [comment]
Agree. but i still prefer `P (e)` format though... https://3v4l.org/sWptM

**Comment by nicolas-grekas** [comment]
I like Maxime's output

**Comment by nicolas-grekas** [comment]
Alternatively, provide P on hover? (Only when different than e?)

**Comment by ro0NL** [comment]
for me the offset adds more value then a location name... hiding it on CLI sounds ike a bad idea :/ in that case i prefer @ogizanagi version as well. It's just a tiny detail after all.. :)

**Comment by maidmaid** [comment]
Updated:

screenshot from 2017-05-03 18-27-01

**Comment by maidmaid** [comment]
Convinced? I do :)

**Comment by maidmaid** [comment]
Strange errors in tests... Could you help me?

**Comment by ro0NL** [comment]
I think you're applying a timezone convesion (from system value to Europe/Zurich). Meaning the timestamp is going to differ...?

Just a guess :))

**Comment by ro0NL** [comment]
```diff
-DateTime @1493503200 {
-  date: 2017-04-30 00:00:00.000000 +02:00
+DateTime @1493535600 {
+  date: 2017-04-30 09:00:00.000000 +02:00
```

i.e. a 9 hours difference..., on appvoyer the default TZ is America/Los_Angeles (utc-8), Europe/Zurich is utc+1..

there's our 9 hours difference :)

**Comment by ogizanagi** [comment]
Indeed @ro0NL is right. Just use the second `\DateTime::construct()` argument to init the timezone (beware it will be ignore in case the time offset is already part of the provided string, like when providing a timestamp (http://php.net/manual/en/datetime.construct.php#refsect1-datetime.construct-parameters)).

I like the new output 😃 , but I may advocate `e (P)` is probably better, because it allows to copy/paste the string (stopped just right before the first parenthesis) to re-create a `\DateTime` instance from it, without loosing information.
I.e:

```php
>>> new \DateTime('2017-04-30 00:00:00.000000 Europe/Zurich'); # <-- 👍
=> DateTime {#167
     +"date": "2017-04-30 00:00:00.000000",
     +"timezone_type": 3,
     +"timezone": "Europe/Zurich",
   }

>>> new \DateTime('2017-04-30 00:00:00.000000 +02:00'); # <-- loosing the geo info
=> DateTime {#171
     +"date": "2017-04-30 00:00:00.000000",
     +"timezone_type": 1,
     +"timezone": "+02:00",
   }

>>> new \DateTime('2017-04-30 00:00:00.000000 Europe/Zurich (+02:00)'); # <-- doesn't even work
Exception with message 'DateTime::__construct(): Failed to parse time string (2017-04-30 00:00:00.000000 Europe/Zurich (+02:00)) at position 41 ((): Unexpected character'
```

I guess it can be useful sometimes.

**Comment by nicolas-grekas** [comment]
Same opinion on e (P) :)

**Comment by ro0NL** [comment]
Given we have a full date string now... :+1: for e (P). I liked the offset in front if prefixed by a `timezone:` key; the first proposal that is.

**Comment by maidmaid** [comment]
Green tests and `` e (P)`` used now :)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        $prefix = Caster::PREFIX_VIRTUAL;
+
+        $a = array();
+        $a[$prefix.'date'] = new ConstStub(
+            $d->format('Y-m-d H:i:s.u '.($d->getTimeZone()->getLocation() ? 'e (P)' : 'P')),
+            sprintf(
+                "literal: %s\nΔnow: %s",
+                $d->format('l, j F Y'),
+                (new \DateTime())->diff($d)->format('%R %yy %mm %dd %H:%I:%S')
```
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: https://github.com/symfony/symfony/pull/22431#discussion_r111608243 :)

**Comment by ro0NL** [comment]
May i do one last suggestion; in the tooltip inform about DST yes/no

https://3v4l.org/BaOcs

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 5 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: I'am rather favor of https://github.com/symfony/symfony/pull/22431#discussion_r114303969

Explanation: if it's 12:00 in Mali (+00:00) and 14:00 in Zurich (+02:00), it's the same moment in the **T**ime, there is no interval of time between. But if we set Zurich's tz, we will have 2h of interval:

```php
>>> (new \DateTime('12:00+00:00'))->diff(new \DateTime('14:00+02:00'))->format('%R %yy %mm %dd %H:%I:%S')
=> "+ 0y 0m 0d 00:00:00" // Ok

>>> (new \DateTime('12:00+02:00'))->diff(new \DateTime('14:00+02:00'))->format('%R %yy %mm %dd %H:%I:%S')
=> "+ 0y 0m 0d 02:00:00" // Ko
```

**Comment by maidmaid** [comment]
Added DST info:
screenshot from 2017-05-05 11-11-24

**Comment by nicolas-grekas** [comment]
Just a few suggestions here: https://gist.github.com/nicolas-grekas/d645a474490b9b00a4f9a55b420ff8d2

- display DST only for geographical times
- tweak the title: remove "literal" and replace Δnow prefix by "from now" suffix
- remove leading zeros from "from now"
- add a test case for the title
- (fixes a bug in StubCaster found while doing so - separate PR coming)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 5 above)
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: You're right!

**Comment by maidmaid** [comment]
Patched (thank you @nicolas-grekas):
screenshot from 2017-05-05 17-48-24

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+use Symfony\Component\VarDumper\Cloner\Stub;
+
+/**
+ * Casts DateTimeInterface related classes to array representation.
+ *
+ * @author Dany Maillard <danymaillard93b@gmail.com>
+ */
+class DateCaster
+{
+    public static function castDate(\DateTimeInterface $d, array $a, Stub $stub, $isNested, $filter)
```
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: `castDateTime` (same for tests)?

**Review by ro0NL — APPROVED** [review]
Awesome!

**Comment by ro0NL** [comment]
Not sure about the literal format though.. shouldnt this show `Saturday, April 3, 2021`? ie. US standard.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 6 above)
Comment (maidmaid) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: In fact, it casts all DateTimeInterface implementations. It's why I choosed the generic castDate method name. castDateTime seems only cast DateTime.

**Comment by maidmaid** [comment]
Rigth about literal format. I will change that.

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 6 above)
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: i'd qualify all those variants a datetime value :) but that's my pov; fine either way :)

**Comment by robfrawley** [comment]
The US standard is the worst... and this is coming from a US citizen. Shouldn't we use what programmers prefer, which is what was already used, ie: `[day] [month] [year]`?

**Comment by ro0NL** [comment]
Not sure what programmers prefer.. :) tbh in programming im quite used to US standards.

Note im fine either way.. i just mentioned it because i believe SF (and PHP) follow this locale in the codebase.

**Comment by maidmaid** [comment]
More, the plus-value of literal format is to show day of week and named month. With U.S. standard, these 2 items are in the front.

**Comment by FabienPapet** [comment]
Shouldn't the 'from now' be 'from server time' , using 'now' can be confusing if the server time is not the developer time

WDYT ?

**Comment by maidmaid** [comment]
@FabienPapet I think that dump a date is in most cases a debug step that is realized on our local machine whose time we known.

**Comment by maidmaid** [comment]
@nicolas-grekas I think this version is quite conclusive. What about merge this and open new PR if we have tweaks?

**Review by nicolas-grekas — APPROVED** [review]
:+1: (To be merged when the feature freeze is over)

**Comment by ogizanagi** [comment]
This should be ready to be merged now :)

File: src/Symfony/Component/VarDumper/Caster/DateCaster.php (hunk 2 above)
Comment (Nek-) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: why waiting?

Comment (fabpot) [inline src/Symfony/Component/VarDumper/Caster/DateCaster.php]: I agree, it would be better to have everything in the same PR.

**Comment by fabpot** [comment]
If we want to support DateInterval, I think it should be done in this PR.

**Comment by maidmaid** [comment]
Ok, I added DateInterval caster.

| Before        | After           |
| ------------- | ------------- |
| screenshot from 2017-06-28 14-17-40 | screenshot from 2017-06-28 14-17-54 |

**Comment by nicolas-grekas** [comment]
Hum, let's remove the DateInterval caster, at least from this PR. The issue I have is accuracy and teaching. Dump should be informational about how things actually works or can be used. On a DateInterval, the listed values are actual public properties. VarDumper is about showing actual facts about objects, not alternate ones ;)
The issue does not exist with DateTime objects because there is no properties there.

BTW, I agree with the comment on the name of the caster method: should be `castDateTime` also IMHO.

**Comment by maidmaid** [comment]
Ok, I removed DateInterval caster and I renamed caster method to ``castDateTime``.

**Comment by nicolas-grekas** [comment]
Thank you @maidmaid.

**Pull request merged**